x-pack/plugins/endpoint/public/applications/endpoint/store/policy_list/test_mock_utils.ts

```diff
@@ -0,0 +1,153 @@
+   * initialized (ex. via `createStore()`. Attempting to reference this property before that time
+   * will throw an error.
+   * Also - do not hold on to references to this property value if `jest.clearAllMocks()` or
+   * `jest.resetAllMocks()` is called between usages of the value.
+   */
+  dispatchSpy: jest.Mock<Dispatch<AppAction>>['mock'];
+  /**
+   * Redux middleware that enables spying on the action that are dispatched through the store
+   */
+  actionSpyMiddleware: ReturnType<MiddlewareFactory<S>>;
+}
+
+/**
+ * Creates a new instance of middleware action helpers
+ * Note: in most cases (testing concern specific middleware) this function should be given
+ * the state type definition, else, the global state will be used.
+ *
+ * @example
+ * // Use in Policy List middleware testing
+ * const middlewareSpyUtils = createSpyMiddleware<PolicyListState>();
+ * store = createStore(
+ *    policyListReducer,
+ *    applyMiddleware(
+ *      policyListMiddlewareFactory(fakeCoreStart, depsStart),
+ *      middlewareSpyUtils.actionSpyMiddleware
+ *    )
+ * );
+ * // Reference `dispatchSpy` ONLY after creating the store that includes `actionSpyMiddleware`
+ * const { waitForAction, dispatchSpy } = middlewareSpyUtils;
+ * //
+ * // later in test
+ * //
+ * it('...', async () => {
+ *   //...
+ *   await waitForAction('serverReturnedPolicyListData');
+ *   // do assertions
+ *   // or check how action was called
+ *   expect(dispatchSpy.calls.length).toBe(2)
+ * });
+ */
+export const createSpyMiddleware = <S = GlobalState>(): MiddlewareActionSpyHelper<S> => {
+  type ActionWatcher = (action: AppAction) => void;
+
+  const watchers = new Set<ActionWatcher>();
+  let spyDispatch: jest.Mock<Dispatch<AppAction>>;
+
+  return {
+    waitForAction: async (actionType: string) => {
+      // Error is defined here so that we get a better stack trace that points to the test from where it was used
+      const err = new Error(`action '${actionType}' was not dispatched within the allocated time`);
+
+      await new Promise((resolve, reject) => {
+        const watch: ActionWatcher = action => {
+          if (action.type === actionType) {
+            watchers.delete(watch);
+            clearTimeout(timeout);
+            resolve();
+          }
+        };
+
+        // We timeout before jest's default 5s, so that a better error stack is returned
+        const timeout = setTimeout(() => {
+          watchers.delete(watch);
+          reject(err);
+        }, 4500);
+        watchers.add(watch);
+      });
+    },
+
+    get dispatchSpy() {
+      if (!spyDispatch) {
+        throw new Error(
+          'Spy Middleware has not been initialized. Access this property only after using `actionSpyMiddleware` in a redux store'
+        );
+      }
+      return spyDispatch.mock;
+    },
+
+    actionSpyMiddleware: api => {
+      return next => {
+        spyDispatch = jest.fn(action => {
```

Comment:
could you create `jest.fn` [where it is defined](https://github.com/elastic/kibana/pull/63291/files#diff-d74b3c93857ebd7176a79896aceeadc8R103)  and here do 

```
spyDispatch.mockImplementation(action => {
  // ...
})
```
would that allow you to remove the comment about an error occurring if you access the mock calls out of sequence?

Reply:
That was my initial thought (in order to avoid the check in the `dispatchSpy` property getter), but dispatch needs access to `next()` thus it needs to be defined within this context.